examples/playlist/raw.go

```diff
@@ -0,0 +1,163 @@
+				"TODO: verify that k8s name and spec.uid match",
+			},
+		},
+		{
+			Version:         "v1-0",
+			SoftwareVersion: "v10.5", // when we remove internal id support
+			Changelog: []string{
+				"removed the dashboard_by_id enumeration type",
+				"removed the PlaylistItem.title property",
+				"remove the spec.uid property",
+				"TODO! added xxx so thema will detect a breaking version change",
+			},
+		},
+	}
+}
+
+// NOTE: this files are not used to do validation, but can be used generically to describe the kind
+func (k *rawPlaylistKind) GetJSONSchema(version string) (string, error) {
+	s, err := packageFS.ReadFile("schemas/" + version + ".json")
+	if err != nil {
+		return "", fmt.Errorf("unknown schema version")
+	}
+	return string(s), err
+}
+
+type ResourceV0 = kindsys.GenericResource[v0x.Spec, kindsys.SimpleCustomMetadata, any]
+type ResourceV1 = kindsys.GenericResource[v1x.Spec, kindsys.SimpleCustomMetadata, any]
+
+func (k *rawPlaylistKind) Read(reader io.Reader, strict bool) (kindsys.Resource, error) {
+	major := -1
+	minor := -1
+
+	var rV0 *ResourceV0
+	var rV1 *ResourceV1
+	var obj kindsys.Resource
+
+	var static kindsys.StaticMetadata
+	var common kindsys.CommonMetadata
+
+	err := kindsys.ReadResourceJSON(reader, kindsys.JSONResourceBuilder{
+		SetGroupVersionKind: func(group, version, kind string) error {
+			if group != k.GetKindInfo().Group {
+				return fmt.Errorf("invalid group")
+			}
+			if kind != k.GetKindInfo().Kind {
+				return fmt.Errorf("invalid kind")
+			}
+			n, err := fmt.Sscanf(version,
+				"v%d-%d", &major, &minor)
+			if err != nil || n != 2 {
+				return fmt.Errorf("unable to read version")
+			}
+			return nil
+		},
+		SetMetadata: func(s kindsys.StaticMetadata, c kindsys.CommonMetadata) {
+			static = s
+			common = c
+		},
+		ReadSpec: func(iter *jsoniter.Iterator) error {
+			if major == 1 {
+				rV1 = &ResourceV1{StaticMeta: static, CommonMeta: common}
+				iter.ReadVal(&rV1.Spec)
+				if iter.Error == nil && strict {
+					err := rV1.Spec.Validate(major, minor)
+					if err != nil {
+						return err
+					}
+				}
+				obj = rV1
+			} else if major == 0 {
+				rV0 = &ResourceV0{StaticMeta: static, CommonMeta: common}
+				iter.ReadVal(&rV0.Spec)
+				if iter.Error == nil && strict {
+					err := rV0.Spec.Validate(major, minor)
+					if err != nil {
+						return err
+					}
+
+					if minor > 0 && rV0.Spec.Uid != rV0.StaticMeta.Name {
+						return fmt.Errorf("the spec.uid must match metadata.name")
+					}
+				}
+				obj = rV0
+			} else {
+				return fmt.Errorf("unknown major version")
+			}
+			return nil
+		},
+		SetAnnotation: func(key, val string) {
+			fmt.Printf("??? unknown")
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	if obj == nil {
+		return nil, fmt.Errorf("missing spec")
+	}
+	return obj, err
+}
+
+func (k *rawPlaylistKind) Migrate(ctx context.Context, obj kindsys.Resource, targetVersion string) (kindsys.Resource, error) {
+	return k.migrator(ctx, obj, targetVersion)
+}
```

Comment:
Hmm, could we define an interface or convention for how migrations should be written and structured? Or is that out of scope for this PR?

Reply:
Afaik, the aim is to leave that to kind/service owners, not forced from Kindsys. Right @ryantxu?

Reply:
yes -- interface in ResourceMigrator

> how migrations should be written and structured? Or is that out of scope for this PR?

yes out of scope here, but I think totally possible for us to explore declarative ways to write migrations.  perhaps based on [JSONPath](https://jsonpath.com/), [JSONata](https://jsonata.org/), or [cel-go](https://github.com/google/cel-go)... or even js+wasm 😝   In general, this feels like something that a single approach would be hard to meet all needs

Reply:
> yes out of scope here [...]

that will be implementation specific, at least for now, right? 🤔